components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/internal/action/PreUpdateProfileActionExecutor.java

```diff
@@ -0,0 +1,304 @@
+                    actionExecutorService.execute(ActionType.PRE_UPDATE_PROFILE, flowContext,
+                            IdentityContext.getThreadLocalIdentityContext().getTenantDomain());
+
+            handleActionExecutionStatus(actionExecutionStatus);
+        } catch (ActionExecutionException e) {
+            throw new UserActionExecutionServerException(UserActionError.PRE_UPDATE_PROFILE_ACTION_SERVER_ERROR,
+                    "Error while executing pre update profile action.", e);
+        }
+    }
+
+    private UserActionRequestDTO buildUserActionRequestDTO(User user, Map<String, String> userClaimsToBeModified,
+                                                           Map<String, String> userClaimsToBeDeleted)
+            throws UserStoreException {
+
+        UserActionRequestDTO.Builder userActionRequestDTOBuilder = new UserActionRequestDTO.Builder()
+                .userId(user.getId());
+
+        populateUpdatingClaimsInUserActionRequestDTO(userClaimsToBeModified, userActionRequestDTOBuilder);
+        populateDeletingClaimsInUserActionRequestDTO(userClaimsToBeDeleted, userActionRequestDTOBuilder);
+
+        return userActionRequestDTOBuilder.build();
+    }
+
+    private void handleActionExecutionStatus(ActionExecutionStatus<?> actionExecutionStatus)
+            throws UserStoreException {
+
+        switch (actionExecutionStatus.getStatus()) {
+            case SUCCESS:
+                return;
+            case FAILED:
+                Failure failure = (Failure) actionExecutionStatus.getResponse();
+                throw new UserActionExecutionClientException(UserActionError.PRE_UPDATE_PROFILE_ACTION_EXECUTION_FAILED,
+                        failure.getFailureReason(), failure.getFailureDescription());
+            case ERROR:
+                Error error = (Error) actionExecutionStatus.getResponse();
+                throw new UserActionExecutionServerException(UserActionError.PRE_UPDATE_PROFILE_ACTION_EXECUTION_ERROR,
+                        error.getErrorMessage(), error.getErrorDescription());
+            default:
+                throw new UserActionExecutionServerException(UserActionError.PRE_UPDATE_PROFILE_ACTION_SERVER_ERROR,
+                        "Unknown execution status received when executing pre update profile action.");
+        }
+    }
+
+    private void populateDeletingClaimsInUserActionRequestDTO(Map<String, String> userClaimsToBeDeleted,
+                                                              UserActionRequestDTO.Builder userActionRequestDTOBuilder)
+            throws UserStoreException {
+
+        for (Map.Entry<String, String> entry : userClaimsToBeDeleted.entrySet()) {
+            String claimKey = entry.getKey();
+            if (isMultiValuedClaim(claimKey)) {
+                userActionRequestDTOBuilder.addClaim(claimKey, new String[]{});
+            } else {
+                userActionRequestDTOBuilder.addClaim(claimKey, "");
```

Comment:
Is this should be empty or null?

Reply:
This is set to an empty string as a mechanism to communicate deleting claims to the external service.

`{
          "uri": "http://wso2.org/claims/emailaddress",
          "value": ""
        },`